addons/knowledge/models/knowledge_article_member.py

```diff
@@ -0,0 +1,108 @@
+        Since this constraint only triggers if we have at least one member on the article, another validation is done in
+        'knowledge.article' model. The article_permission related field has been added and stored to force triggering
+        this constraint when article.permission is modified.
+        :param on_unlink: (Boolean) When called on unlink, we must remove the members in self
+            (the ones that will be deleted) to check if one of the remaining members has write access.
+        """
+        articles_to_check = self.article_id.filtered(lambda a: a.inherited_permission != 'write')
+        if not articles_to_check:
+            return
+
+        if on_unlink:
+            deleted_members_by_article = dict.fromkeys(articles_to_check.ids, self.env['knowledge.article.member'])
+            for member in self.filtered(lambda member: member.article_id in articles_to_check):
+                deleted_members_by_article[member.article_id.id] |= member
+
+        parents_members_permission = articles_to_check.parent_id._get_article_member_permissions()
+        for article in articles_to_check:
+            # Check on permission on members
+            members_to_check = article.article_member_ids
+            if on_unlink:
+                members_to_check -= deleted_members_by_article[article.id]
+            if any(m.permission == 'write' for m in members_to_check):
+                continue
+
+            # we need to add the members on parents to check the validity
+            parent_write_members = any(
+                values['permission'] == 'write' for partner_id, values
+                in parents_members_permission[article.parent_id.id].items()
+                if article.parent_id and not article.is_desynchronized
+                and partner_id not in article.article_member_ids.partner_id.ids
+            )
+
+            if not parent_write_members:
+                raise ValidationError(
+                    _("Article '%s' should always be available for update: inherit write permission, or have a member with write access)",
+                      article.display_name)
+                )
+
+    @api.constrains('partner_id', 'permission')
+    def _check_external_member_permission(self):
+        for member in self:
+            if member.partner_id.partner_share and member.permission == 'write':
+                raise ValidationError(_("The external user %s cannot have a 'write' permission on article '%s'", member.partner_id.display_name, member.article_id.display_name))
+
+    @api.depends("article_id", "permission")
+    def _compute_has_higher_permission(self):
+        articles_permission = self.article_id._get_internal_permission()
+        for member in self:
+            member.has_higher_permission = ARTICLE_PERMISSION_LEVEL[member.permission] > ARTICLE_PERMISSION_LEVEL[articles_permission[member.article_id.ids[0]]]
```

Comment:
```suggestion
            member.has_higher_permission = ARTICLE_PERMISSION_LEVEL[member.permission] > ARTICLE_PERMISSION_LEVEL[articles_permission[member.article_id.id]]
```

Reply:
@fbarbe00

Reply:
> so if member has `read` and the article level is `none`, we consider the member can administrate or I am missing something?

luckily not, this is only used to prevent the user to give itself a higher permission when removing themselves in `_remove_member`. There is currently a security issue there (which allows you to delete members when you sometimes can't), but @tde-banana-odoo or @dbeguin should be on it

Reply:
Indeed there was a privilege escalation issue. Then there was a privilege removal issue. This should be fixed now, we have some unit tests covering those use cases. We lessened use of sudo() and added some protections here and there.